src/main/java/org/apache/pulsar/io/jcloud/batch/BatchContainer.java

```diff
@@ -37,21 +40,35 @@ public class BatchContainer {
     private final long maxBatchTimeMs;
     private final AtomicLong currentBatchSize = new AtomicLong(0L);
     private final AtomicLong currentBatchBytes = new AtomicLong(0L);
-    private final ArrayBlockingQueue<Record<GenericRecord>> pendingFlushQueue;
-    private volatile long lastPoolRecordsTime;
+    private volatile long lastPollRecordsTime;
+    private final List<Record<GenericRecord>> pendingFlushList;
+    private final ReentrantLock lock = new ReentrantLock();
+    private final Condition notFull = lock.newCondition();
 
-    public BatchContainer(long maxBatchSize, long maxBatchBytes, long maxBatchTimeMs, int maxPendingQueueSize) {
+    public BatchContainer(long maxBatchSize, long maxBatchBytes, long maxBatchTimeMs) {
         this.maxBatchSize = maxBatchSize;
         this.maxBatchBytes = maxBatchBytes;
         this.maxBatchTimeMs = maxBatchTimeMs;
-        this.pendingFlushQueue = new ArrayBlockingQueue<>(maxPendingQueueSize);
-        this.lastPoolRecordsTime = System.currentTimeMillis();
+        this.lastPollRecordsTime = System.currentTimeMillis();
+        this.pendingFlushList = new LinkedList<>();
     }
 
     public void add(Record<GenericRecord> record) throws InterruptedException {
-        pendingFlushQueue.put(record);
-        updateCurrentBatchSize(1);
-        updateCurrentBatchBytes(record.getMessage().get().size());
+        lock.lock();
+        try {
+            // Allow exceeding the maximum value once
+            long recordSize = record.getMessage().get().size();
+            pendingFlushList.add(record);
+            currentBatchSize.incrementAndGet();
+            currentBatchBytes.addAndGet(recordSize);
+
+            // Wait if the batch needs to be flushed
+            while (needFlush()) {
+                notFull.await();
+            }
```

Comment:
Can we schedule the flush immediately here and so that we don't nee the `notFull`?

Reply:
This class does not handle flush.